pkg/config/apiserver.go

```diff
@@ -0,0 +1,210 @@
+	c := restful.NewContainer()
+	a := &API{
+		version:  version,
+		rootPath: fmt.Sprintf("/api/%s", version),
+		store:    store,
+		readBody: ioutil.ReadAll,
+		validate: func(cfg map[string]string) (rt *Validated, ce *adapter.ConfigErrors) {
+			v := newValidator(aspectFinder, builderFinder, findAspects, true, eval)
+			return v.validate(cfg)
+		},
+	}
+	a.register(c)
+	a.Server = &http.Server{Addr: ":" + strconv.Itoa(port), Handler: c}
+	a.handler = c
+	return a
+}
+
+// Run calls listen and serve on the API server
+func (a *API) Run() {
+	glog.Infof("Starting Config API Server at %v", a.Server.Addr)
+	glog.Warning(a.Server.ListenAndServe())
+}
+
+// getRules returns the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) getRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+	val, index, found := a.store.Get(funcPath)
+	if !found {
+		writeResponse(http.StatusNotFound, fmt.Sprintf("no rules for %s\n", funcPath), resp)
+		return
+	}
+
+	// TODO send index back to the client
+	_ = index
+	resp.AddHeader("Content-Type", "application/yaml")
+	write(val, resp)
+}
+
+// putRules replaces the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) putRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+
+	// TODO optimize only read descriptors and adapters
+	data, _, index, err := readdb(a.store, "/")
+	if err != nil {
+		writeResponse(http.StatusInternalServerError, err.Error(), resp)
+		return
+	}
+	// TODO send index back to the client
+	_ = index
+
+	var bval []byte
+	bval, err = a.readBody(req.Request.Body)
+	if err != nil {
+		writeResponse(http.StatusInternalServerError, err.Error(), resp)
+		return
+	}
+	val := string(bval)
+	data[funcPath] = val
+
+	_, cerr := a.validate(data)
+	if cerr != nil {
+		glog.Warning(cerr.Error())
+		glog.Warning(val)
+		writeResponse(http.StatusPreconditionFailed, cerr.Error(), resp)
+		return
+	}
+	index, err = a.store.Set(funcPath, val)
+	if err != nil {
+		writeResponse(http.StatusInternalServerError, err.Error(), resp)
+		return
+	}
+	// TODO send index back to the client
+	_ = index
+
+	writeResponse(http.StatusOK, fmt.Sprintf("Created %s", funcPath), resp)
+}
+
+// a subset of restful.Response
+type response interface {
+	// WriteHeaderAndJson is a convenience method for writing the status and a value in Json with a given Content-Type.
+	WriteHeaderAndJson(status int, value interface{}, contentType string) error
+}
+
+func write(contents string, resp io.Writer) {
+	_, err := resp.Write([]byte(contents))
+	if err != nil {
+		glog.Warning(err)
+	}
+}
+
+func writeResponse(httpStatus int, msg string, resp response) {
+	if err := resp.WriteHeaderAndJson(
+		httpStatus,
+		status.WithMessage(
```

Comment:
it would be sweet if we could generate status messages that included appropriate error details, including things like DebugInfo (https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L47) and BadRequest (https://github.com/googleapis/googleapis/blob/master/google/rpc/error_details.proto#L91).